### PR TITLE
KAFKA-14653: MirrorMakerConfig using raw properties instead of post-r…

### PR DESCRIPTION
**Pull request by urbandan**
…esolution properties.

MirrorMakerConfig extends AbstractConfig, which resolves config provider based references eagerly, construction time. Because of this, Connector configurations created by MirrorMaker2 always contain resolved property values, and never provider references. This can cause security, and host-specific configuration issues.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by urbandan** [comment]
@C0urante @viktorsomogyi Can you please take a look?

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java
Comment (C0urante) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: This counts as a change to public interface since subclasses of `AbstractConfig` would be able to access this new method. And since this wasn't mentioned in the KIP, we probably shouldn't do that here.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java
Comment (C0urante) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java]: It looks like this is intended to get around the fact that the constructor accepts a `Map<?, ?>` but we really want to work with a `Map<String, String>`. Could we change the constructor to accept a `Map<String, String>` and store that as our `rawProperties` field, instead of introducing the `copyAsStrings` method?

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java
Comment (C0urante) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java]: Why change this part? Aren't we transforming the configs later at line 141 anyways?

**Review by C0urante — COMMENTED** [review]
Thanks @urbandan, left a few thoughts. Going to have to go through this part of the codebase more carefully before approving but wanted to at least get out a first round today.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java
Comment (urbandan) [reply clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: Originally, I had this inlined in the MMConfig ctor, but didn't want to duplicate code - anyway, I'll inline it again, that's safer

**Comment by urbandan** [comment]
@C0urante Thanks for the comments, addressed them

**Review by C0urante — COMMENTED** [review]
Thanks @urbandan! I've given the functional parts a more thorough passthrough this time and everything seems good--we're ensuring that all properties used for connector configs are derived from the raw, pre-transformation properties provided by the user, and that all other properties for, e.g., worker configs, are still transformed before being used.

I have one final request before we merge this: can we flesh out the `testLazyConfigResolution` test case with some assertions that properties that should be transformed are still transformed? Ideally we could do this for every public-facing method on the `MirrorMakerConfig` class, but realistically I think it'd be fine if we just settled for the `clientConfig` and `workerConfig` methods.

**Comment by urbandan** [comment]
@C0urante Existing tests (testWorkerConfigs and testClientConfigProperties) already cover the config provider resolution (they use ${fake: secret:password} and FakeConfigProvider), I don't think there is much else to add in this regard.

As an additional note, the explicit transformation in MirrorMakerConfig is an unnecessary step, as both client (MirrorClientConfig) and worker configs (DistributedConfig) get all the config provider configurations, and since those classes extend AbstractConfig, they also perform the config resolution step. Should we remove the explicit transform from MirrorMakerConfig?

**Comment by urbandan** [comment]
Rebased the change on trunk and resolved conflicts

**Comment by C0urante** [comment]
@urbandan Ah yes, apologies for missing that. You're right, we don't need any more coverage for the `MirrorMakerConfig` class.

I also agree that the manual config transformations performed by that class appear to be unnecessary. I'm a little hesitant to remove them since I haven't been able to find out why they were added that way in the first place. I checked the original MM2 KIP and PR, but couldn't find anything relevant, and the `ConfigProvider` mechanism was added to the `AbstractConfig` class in version 2.3.0, one release earlier than MM2, which came out in 2.4.0. I guess it's possible that the PR for MM2 was written before the `ConfigProvider` logic was added to the `AbstractConfig` class, and the manual transformation slipped through during review?

Anyways, I think it'd be best to leave it to a follow-up PR if we wanted to remove that logic. That way, if there is any unintended fallout, we can just revert that commit without worrying about also undoing the work necessary for KIP-710.

**Review by C0urante — APPROVED** [review]
LGTM, thanks @urbandan! Will merge pending CI build.

**Comment by urbandan** [comment]
Thanks for the reviews @C0urante!

**Comment by C0urante** [comment]
Test failures appear unrelated (the newly-introduced dedicated MM2 integration tests seem to be flaking out on Jenkins; filed KAFKA-14718 to track). Merging...

**Pull request merged**
